### DISCUSSION TITLE
how to run the inference in pure C code?

### DISCUSSION THREAD

**Discussion by ardeal** [microsoft/onnxruntime#19740] [API Q&A] [answered]
I have already trained yolov8 in python code, and have already converted the pth model to onnx model.
I would like to run the inference of the onnx mode in pure C code, how to run it? is there any example code, workflow or document?

**Comment by pranavsharma** [comment, marked as answer]
https://github.com/microsoft/onnxruntime-inference-examples/tree/main/c_cxx/fns_candy_style_transfer

**Reply by ardeal** [reply]
Thank you @pranavsharma !

in your examples, you are using exe to run inference.
what I would like to do are:
1) pure C code inference.
2) run on visual studio and embedded MCU respectively.
so, I would like to check the pure c code and examples about how to run it.

**Reply by pranavsharma** [reply]
The pure C code is in the same folder in file https://github.com/microsoft/onnxruntime-inference-examples/blob/main/c_cxx/fns_candy_style_transfer/fns_candy_style_transfer.c. You can run cmake using the example CMakeLists.txt file and it'll generate a platform dependent executable that you can run.

**Reply by ardeal** [reply]
Thank you @pranavsharma ,
The C code in `fns_candy_style_transfer.c` is not what I am looking for.

I would like to check the operators(i.e. conv2d, resize, linear and etc) C code. 

according to the CMakeList.txt file, I guess the C code is in the library **onnxruntime**.  However, I could not fiind the operators  C code in onnxruntime after I checked onnxruntime in repo: https://github.com/microsoft/onnxruntime. 

I would like to copy the operators C code to embedded MCU and call them with similar C code like `fns_candy_style_transfer.c`

![image](https://github.com/microsoft/onnxruntime/assets/6227348/3c43d4fd-5eed-4920-bd20-9bf91c4658b6)

**Reply by pranavsharma** [reply]
Code for all the operators are here https://github.com/microsoft/onnxruntime/tree/main/onnxruntime/core/providers/cpu.

**Comment by ardeal** [comment]
Than you @pranavsharma !

The code in https://github.com/microsoft/onnxruntime/tree/main/onnxruntime/core/providers/cpu is C++ code, but not C code. 
What I am looking for is C code.

**Reply by pranavsharma** [reply]
Sorry, this is a C++ project. We don't have C code.

**Reply by ardeal** [reply]
in `onnxruntime-inference-examples`,  `fns_candy_style_transfer.exe <model_path> <input_image_path> <output_image_path> [cpu|cuda|dml]` is a C file, right?

The upper exe will call `onnxruntime` lib which is a C++. 
This confused me.  Could the C exe call the C++ lib?

**Reply by pranavsharma** [reply]
An executable links with the ORT .so/DLL (implicitly [at link time] or explicitly [by calling dlopen]) and uses the C APIs exported by it. Do you have a concrete use case?

**Reply by ardeal** [reply]
Thank you @pranavsharma !

I really cannot understand. your operators are C++ code, while the exported DLL/APIs are C.  it means that you are using C to call C++ functions in the dll. is my understanding correct?

**Reply by pranavsharma** [reply]
Please familiarize yourself with how shared libs work. This is not an ORT discussion/issue.
